### PR TITLE
Implement ruriv3 reranker

### PR DESCRIPTION
**Pull request by ooe1123**
https://github.com/ailia-ai/ailia-models/issues/1698

### CODE REVIEW

**Comment by ooe1123** [comment]
ailia_tokenizerだとスコアが変わってしまうようで、無効にしています

**Comment by kyakuno** [comment]
モデルアップロード済み。
https://storage.googleapis.com/ailia-models/ruri-v3-reranker/ruri-v3-reranker-310m.onnx.prototxt
など

**Comment by kyakuno** [comment]
padが0か3かが異なる。

disable_ailia_tokenizer

```
[[    1 64260  1200 18617  1200   368     1 46632   275 17738   275 26315
  46632   275 49294  2575  1918   275  2862  3781   275 63308   306  4746
    893   685   278  4510 15895   246   181   204 43353  4141  2587 28359
   3220   297   277   303  1347   277   300 14649   293   471 34428  4452
    312 20937  4452   312 56987  4452  1942   245   177   174  4452 56754
    824   278  2518  4141  2587   302 21688  2046  4718   315 44355 23612
    297   343  1058   293   320  2004   275 39880 10113   297  2862  3781
    275 26315 46632   293   275   454   378   735   297 46632   275 17738
    275 49294  2575  1918   453   293   275   280 68932   297 63308   293
    275  5757  8254   297 91509   293   291   294 17021 20631  1184   278]
 [    1 64260  1200 18617  1200   368     1 64260   275  1645 79135 62092
    321   516   756   327   302   275 20851   291  4131  7874 11211   278
   2698 14866 73022   291 14942   382   240   199   156 44793  1785   351
    271  1645   501 98180 73022  2123 18178  5349   685   278   777  5969
   1830  3499  2597   275  1410  5383 26082  1206  1265   685 81257   376
    277   379   278     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3]
 [    1 64260  1200 18617  1200   368     1 64260  1200   321   516   756
  26933   327   302   275 10325  1214   315  5479 67780 13339  2925   278
  21712   275  4856  5837 36237 64260   321  1410  5383 26082  1206  1265
    275  4266   287 30365   654   890 10476   782   327   832   278 38765
  45012   783  1200   857   467   313 24116 10325 28209  2925  2642 14730
   1405   271 11359   284   430   473   327   317 13869   732   376   277
   2244   279   379   278     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3
      3     3     3     3     3     3     3     3     3     3     3     3]]
```

ailia_tokenizer

```
[[    1 64260  1200 18617  1200   368     1 46632   275 17738   275 26315
  46632   275 49294  2575  1918   275  2862  3781   275 63308   306  4746
    893   685   278  4510 15895   246   181   204 43353  4141  2587 28359
   3220   297   277   303  1347   277   300 14649   293   471 34428  4452
    312 20937  4452   312 56987  4452  1942   245   177   174  4452 56754
    824   278  2518  4141  2587   302 21688  2046  4718   315 44355 23612
    297   343  1058   293   320  2004   275 39880 10113   297  2862  3781
    275 26315 46632   293   275   454   378   735   297 46632   275 17738
    275 49294  2575  1918   453   293   275   280 68932   297 63308   293
    275  5757  8254   297 91509   293   291   294 17021 20631  1184   278]
 [    1 64260  1200 18617  1200   368     1 64260   275  1645 79135 62092
    321   516   756   327   302   275 20851   291  4131  7874 11211   278
   2698 14866 73022   291 14942   382   240   199   156 44793  1785   351
    271  1645   501 98180 73022  2123 18178  5349   685   278   777  5969
   1830  3499  2597   275  1410  5383 26082  1206  1265   685 81257   376
    277   379   278     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0]
 [    1 64260  1200 18617  1200   368     1 64260  1200   321   516   756
  26933   327   302   275 10325  1214   315  5479 67780 13339  2925   278
  21712   275  4856  5837 36237 64260   321  1410  5383 26082  1206  1265
    275  4266   287 30365   654   890 10476   782   327   832   278 38765
  45012   783  1200   857   467   313 24116 10325 28209  2925  2642 14730
   1405   271 11359   284   430   473   327   317 13869   732   376   277
   2244   279   379   278     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0
      0     0     0     0     0     0     0     0     0     0     0     0]]
```

**Comment by kyakuno** [comment]
ただ、attention_maskがあるので、どっちでも数値は合う。

**Comment by kyakuno** [comment]
ailia_tokenizerを使っても使わなくても、[2]が0.968になる。

```
[3] (1.000)
- 瑠璃色（るりいろ）は、紫みを帯びた濃い青。名は、半貴石の瑠璃（ラピスラズリ、英: lapis lazuli）による。JIS慣用色名では「こい紫みの青」（略号 dp-pB）と定義している[1][2]。

[2] (0.968)
- 瑠璃、または琉璃（るり）は、仏教の七宝の一つ。サンスクリットの vaiḍūrya またはそのプラークリット形の音訳である。金緑石のこととも、ラピスラズリであるともいう[1]。

[1] (0.001)
- ワシ、タカ、ハゲワシ、ハヤブサ、コンドル、フクロウが代表的である。これらの猛禽類はリンネ前後の時代(17~18世紀)には鷲類・鷹類・隼類及び梟類に分類された。ちなみにリンネは狩りをする鳥を単一の目(もく)にまとめ、vultur(コンドル、ハゲワシ)、falco(ワシ、タカ、ハヤブサなど)、strix(フクロウ)、lanius(モズ)の4属を含めている。
```

**Comment by kyakuno** [comment]
ailiaもortも同じ結果。

**Comment by kyakuno** [comment]
@ooe1123 --diable_ailia_tokenizerの有無で結果が変わりますでしょうか？readmeを見ると、2番目が(0.056)が期待値のところ、(0.968)になっているのが問題だと思うのですが、tokenizerの値は一致していそうなので、そこが問題ではないような気もしています。問題の起きる環境で、input_idsをprintして見て頂いても良いでしょうか？

**Comment by ooe1123** [comment]
@kyakuno

手元で試してみたところ、input_idsが違うようでした。

確認コード
[test_tokenizer_diff.py](https://github.com/user-attachments/files/29137119/test_tokenizer_diff.py)

**Comment by kyakuno** [comment]
元モデル
https://huggingface.co/cl-nagoya/ruri-v3-reranker-310m/tree/main

transformersでは、tokenizer.modelが存在しているが、tokenizer.jsonでBPE指定されているため、実際は使用していない。
ailia_tokenizerでは、tokenizer.modelを使用して、SentencePieceを使用するため、不一致になっている。

**Comment by kyakuno** [comment]
tokenizerフォルダのtokenizer.jsonとtokenizer_config.jsonが正しいファイルではなかった模様。
ファイルを再ダウンロードして更新すると、下記の二つは一致した。

```
tok_hf = PreTrainedTokenizerFast.from_pretrained("cl-nagoya/ruri-v3-reranker-310m")
tok_ai = GemmaTokenizer.from_pretrained("./tokenizer")
```

**Comment by kyakuno** [comment]
ただ、本来はAutoTokenizerとPreTrainedTokenizerFastの結果は一致するはず。
現在は、transformersで、AutoTokenizerとPreTrainedTokenizerFastで結果が異なる。

**Comment by kyakuno** [comment]
LlamaTokenizerFast は tokenizer.json を無視して tokenizer.model から BPEを再構築するのに食い違っている。
tokenizer.modelも間違っている可能性がある。

**Comment by kyakuno** [comment]
tokenizer.modelは正しそう。しかし、下記の2つでも食い違う。

```
tok_hf = AutoTokenizer.from_pretrained("tokenizer") -> ailia_tokenierと異なる
tok_hf = AutoTokenizer.from_pretrained("cl-nagoya/ruri-v3-reranker-310m") -> ailia_tokenizerと同じ
```

**Comment by kyakuno** [comment]
AutoTokenizer はローカルパスの場合 tokenizer_class: LlamaTokenizer を見てLlamaTokenizerFast を強制し、tokenizer.json を無視してしまう。hubからはtokenizer.json を優先する。これは AutoTokenizerのローカル/hub間の挙動の差。AutoTokenizer はローカルパスでは使わず PreTrainedTokenizerFastを使うのが良い模様。

**Comment by kyakuno** [comment]
最終的に、下記にすると、どちらもリファレンスと一致しました。

```
    if args.disable_ailia_tokenizer:
        from transformers import AutoTokenizer
        #tokenizer = AutoTokenizer.from_pretrained("cl-nagoya/ruri-v3-reranker-310m") # リファレンス
        #tokenizer = AutoTokenizer.from_pretrained("tokenizer") # 不一致
        from transformers import PreTrainedTokenizerFast
        tokenizer = PreTrainedTokenizerFast.from_pretrained("tokenizer") # 一致
    else:
        #from ailia_tokenizer import LlamaTokenizer
        #tokenizer = LlamaTokenizer.from_pretrained("./tokenizer") # BOSあり、EOSなし、不一致

        from ailia_tokenizer import GemmaTokenizer
        tokenizer = GemmaTokenizer.from_pretrained("./tokenizer") # BOSあり、EOSあり、一致
```

```
[3] (1.000)
- 瑠璃色（るりいろ）は、紫みを帯びた濃い青。名は、半貴石の瑠璃（ラピスラズリ、英: lapis lazuli）による。JIS慣用色名では「こい紫みの青」（略号 dp-pB）と定義している[1][2]。

[2] (0.081)
- 瑠璃、または琉璃（るり）は、仏教の七宝の一つ。サンスクリットの vaiḍūrya またはそのプラークリット形の音訳である。金緑石のこととも、ラピスラズリであるともいう[1]。

[1] (0.000)
- ワシ、タカ、ハゲワシ、ハヤブサ、コンドル、フクロウが代表的である。これらの猛禽類はリンネ前後の時代(17~18世紀)には鷲類・鷹類・隼類及び梟類に分類された。ちなみにリンネは狩りをする鳥を単一の目(もく)にまとめ、vultur(コンドル、ハゲワシ)、falco(ワシ、タカ、ハヤブサなど)、strix(フクロウ)、lanius(モズ)の4属を含めている。
 INFO ruri-v3-reranker.py (148) : Script finished successfully.
```

**Comment by kyakuno** [comment]
通常のruriv3も一致していない気がする。

**Comment by kyakuno** [comment]
ruriv3も同様に修正。

**Comment by kyakuno** [comment]
ruri-v3-rerankerのjsonを更新した後は、ruri-v3-rerankerとruri-v3のファイルは一致する。

```
kyakuno@MacBookAir-M2-KK tokenizer % diff tokenizer_config.json ../../ruri-v3/tokenizer/tokenizer_config.json 
kyakuno@MacBookAir-M2-KK tokenizer % diff tokenizer.json ../../ruri-v3/tokenizer/tokenizer.json       
kyakuno@MacBookAir-M2-KK tokenizer % diff tokenizer.model ../../ruri-v3/tokenizer/tokenizer.model
kyakuno@MacBookAir-M2-KK tokenizer % diff special_tokens_map.json ../../ruri-v3/tokenizer/special_tokens_map.json
```

**Comment by kyakuno** [comment]
ailia_tokenizerにおいて、LlamaTokenizerもGemmaTokenizerもtokenizer.modelを使用するので、EOSの有無しか結果に影響しない。
transformesにおいては、LlamaTokenizerとGemmaTokenizerは、Normalizer（スペースを_に置き換え）とPre-tokenizer（MetaspaceかSplitか）を含む設計差がある。
これらの設定はSentencePieceのモデルファイルに焼き込まれるので、ailia_tokenizer経由だと影響しない。

**Review by kyakuno — APPROVED** [review]

**Comment by kyakuno** [comment]
transformersは下記で差異が出ている。tokenizer.modelは参照されない。

PreTrainedTokenizerFast -> tokenizer.json を直接ロード → Unigram モデル
AutoTokenizer (LlamaTokenizer) -> cls.model = BPE なので vocab を BPE 辞書として変換 -> BPE + merges 空 = 単文字ごとにトークン化

本質的には、tokenizer_config.json に書かれた "tokenizer_class": "LlamaTokenizer"と、実際の tokenizer.json の中身が矛盾している。ruri-v3 が Llama  系アーキテクチャをベースにしているため tokenizer_class: LlamaTokenizerを使ったが、実際には日本語向けにカスタムの Unigram  トークナイザーに置き換えた際に tokenizer_class を更新し忘れたのが原因。

ruri-v3 の tokenizer_config.json 側のバグと考えられる。 ただし HuggingFace Hub上のファイルを修正する権限はないため、コード側で PreTrainedTokenizerFast を使う現在の対処は正しい回避策。

そもそも、ruri-v3がLlamaTokenizerではないということなので、ailiaがLlamaTokenizerではなく、GemmaTokenizerを使ってもサンプルコードとしておかしくない。

**Comment by kyakuno** [comment]
近いIssueは  #45488 (https://github.com/huggingface/transformers/issues/45488) 。

  LlamaTokenizer v5 が tokenizer.json の設定を上書きする（Open）
  - tokenizer_class: LlamaTokenizerFast だが tokenizer.json が ByteLevel BPE の
  DeepSeek V3/R1 で発生
  - LlamaTokenizer.__init__ が Metaspace pre-tokenizer
  を強制上書きするため無言で壊れる
  - 回避策：PreTrainedTokenizerFast.from_pretrained() — 今回と同じ対処法
